File: src/python/scraper/videos/base/parser.py
from bs4 import BeautifulSoup
import re
from .config import config

def get_videos(soup):
    return [parse_video(video) for video in soup.select('#content > .mozaique > .thumb-block') if parse_video(video) is not None]

def get_pages(soup, is_fresh):
    pages = set()
    for page in soup.select('.pagination > ul > li > a'):
        text = page.text.strip()
        if text.isdigit():
            page_number = int(text)
            pages.add(page_number - 1 if is_fresh else page_number)
    
    last_page = soup.select_one('.last-page')
    if last_page:
        text = last_page.text.strip()
        if text.isdigit():
            last_page_number = int(text)
            pages.add(last_page_number - 1 if is_fresh else last_page_number)

    return sorted(pages)

def parse_response(page, data, is_fresh=False):
    soup = BeautifulSoup(data, 'html.parser')
    videos = get_videos(soup)
    results_count = extract_results_count(soup, is_fresh)
    pages = get_pages(soup, is_fresh)
    last_page = pages[-1]
    
    if not is_fresh:
        last_page = 149 if int(results_count) >= 4029 else (results_count // 27) + 1
    
    pagination = {
        'page': page + 1,
        'pages': pages,
        'last_page': last_page
    }
    
    return {
        'videos': videos,
        'pagination': pagination,
        'has_next': page < last_page,
        'next': page + 1 if page < last_page else None,
        'has_previous': page > min(pages),
        'previous': page - 1 if page > min(pages) else None,
        'results_count': results_count,
    }

def extract_results_count(soup, is_fresh=False):
    results_span = soup.select_one('h2.page-title .sub')
    if results_span:
        results_text = results_span.text.strip()
        match = re.search(r'\(([\d.,]+)', results_text)
        if match:
            results_count = int(match.group(1).replace('.', '').replace(',', ''))
        else:
            results_count = "Unknown"
    else:
        results_count = "Unknown"
    
    if is_fresh:
        results_count = 27 * 20000
    
    return results_count

def extract_title(video):
    title_element = video.select_one('.title a')
    if not title_element:
        raise ValueError("Title element not found")
    return title_element.get('title')

def extract_path_and_url(video):
    path_element = video.select_one('.thumb > a')
    if not path_element:
        raise ValueError("Path element not found")
    path = path_element.get('href')
    url = f"{config.get_base_url()}{path}"
    return path, url

def extract_duration(video):
    duration_element = video.select_one('.metadata .duration')
    return duration_element.text.strip() if duration_element else "Unknown"

def extract_profile(video):
    profile_element = video.select_one('.metadata span a')
    return {
        'name': profile_element.text.strip() if profile_element else "Unknown",
        'url': f"{config.get_base_url()}{profile_element.get('href')}" if profile_element else "Unknown"
    }

def extract_views(video):
    views = "Unknown"
    metadata = video.select_one('.metadata')
    if metadata:
        sprfluous_spans = metadata.find_all('span', class_='sprfluous')
        if len(sprfluous_spans) >= 2:
            views_text = sprfluous_spans[0].find_next_sibling(text=True)
            if views_text:
                views = views_text.strip().split()[0]
    return views

def parse_video(video):
    try:
        title = extract_title(video)
        path, url = extract_path_and_url(video)
        duration = extract_duration(video)
        profile = extract_profile(video)
        views = extract_views(video)
        return {
            'url': url,
            'path': path,
            'title': title,
            'duration': duration,
            'profile': profile,
            'views': views,
        }
    except ValueError as e:
        print(f"Error parsing video: {e}")
        return None

File: src/api/utils/parser.ts
import cheerio from "cheerio";
import config from "../config/config";

class Parser {
  private isFresh: boolean;
  constructor(isFresh: boolean = false) {
    this.isFresh = isFresh;
  }
  public parseResponse(
    page: number,
    data: string,
  ): { videos: Video[]; pagination: Pagination } {
    const $ = cheerio.load(data);
    const videos = this.getVideos($);
    const resultsCount = this.extractResultsCount($);
    const pages = this.getPages($);
    let lastPage = pages.at(-1);
    if (!this.isFresh) {
      lastPage = resultsCount >= 4029 ? 149 : Math.ceil(resultsCount / 27);
    }
    const pagination: Pagination = {
      page: page + 1,
      pages,
      last_page: lastPage,
      has_next: page < lastPage,
      next: page < lastPage ? page + 1 : null,
      has_previous: page > Math.min(...pages),
      previous: page > Math.min(...pages) ? page - 1 : null,
      results_count: resultsCount,
    };
    return { videos, pagination };
  }
  private getVideos($: cheerio.Root): Video[] {
    const videos: Video[] = [];
    $("#content .mozaique .thumb-block").each((_, element) => {
      const video = this.parseVideo($(element));
      if (video) videos.push(video);
    });
    return videos;
  }

  private getPages($: cheerio.Root): number[] {
    const pages = new Set<number>();
    $(".pagination ul li a").each((_, element) => {
      const text = $(element).text().trim();
      if (!isNaN(parseInt(text))) {
        const pageNumber = parseInt(text);
        pages.add(this.isFresh ? pageNumber - 1 : pageNumber);
      }
    });
    return Array.from(pages).sort((a, b) => a - b);
  }

  private extractResultsCount($: cheerio.Root): number {
    if (this.isFresh) return 27 * 20000;
    const resultsSpan = $("h2.page-title .sub");
    if (resultsSpan.length) {
      const resultsText = resultsSpan.text().trim();
      const match = resultsText.match(/\(([\d.,]+)/);
      if (match) {
        return parseInt(match[1].replace(/[.,]/g, ""));
      }
    }
    throw new Error("No results found");
  }
  private extractTitle(video: cheerio.Cheerio): string {
    const titleElement = video.find(".title a");
    if (!titleElement.length) throw new Error("Title element not found");
    return titleElement.attr("title") || "";
  }

  private extractPathAndUrl(video: cheerio.Cheerio): {
    path: string;
    url: string;
  } {
    const pathElement = video.find(".thumb > a");
    if (!pathElement.length) throw new Error("Path element not found");
    const path = pathElement.attr("href") || "";
    const url = `${config.getBaseUrl()}${path}`;
    return { path, url };
  }

  private extractDuration(video: cheerio.Cheerio): string {
    const durationElement = video.find(".metadata .duration");
    return durationElement.length ? durationElement.text().trim() : "Unknown";
  }

  private extractProfile(video: cheerio.Cheerio): {
    name: string;
    url: string;
  } {
    const profileElement = video.find(".metadata span a");
    return {
      name: profileElement.length ? profileElement.text().trim() : "Unknown",
      url: profileElement.length
        ? `${config.getBaseUrl()}${profileElement.attr("href")}`
        : "Unknown",
    };
  }
  private extractViews(video: cheerio.Cheerio): string {
    let views = "Unknown";
    const metadata = video.find(".metadata");
    if (metadata.length) {
      const sprfluousSpans = metadata.find("span.sprfluous");
      if (sprfluousSpans.length >= 2) {
        const viewsText = sprfluousSpans[0].next.data.trim();
        return viewsText;
      }
    }
    return views;
  }

  private parseVideo(video: cheerio.Cheerio): Video | null {
    try {
      const title = this.extractTitle(video);
      const { path, url } = this.extractPathAndUrl(video);
      const duration = this.extractDuration(video);
      const profile = this.extractProfile(video);
      const views = this.extractViews(video);
      return {
        url,
        path,
        title,
        duration,
        profile,
        views,
      };
    } catch (e) {
      return null;
    }
  }
}

export default Parser;
